crates/ra_hir/src/ty/infer.rs

```diff
@@ -54,13 +55,33 @@ pub fn infer(db: &impl HirDatabase, func: Function) -> Arc<InferenceResult> {
     Arc::new(ctx.resolve_all())
 }
 
+#[derive(Debug, Copy, Clone, Hash, PartialEq, Eq)]
+pub enum ExprOrPatId {
+    Expr(ExprId),
+    Pat(PatId),
+}
+
+impl From<ExprId> for ExprOrPatId {
```

Comment:
Let's use the `impl_froms` macro here? I think we'll need to change the name of `enum` variants for this though.